src/main/scala/org/camunda/feel/impl/interpreter/FeelInterpreter.scala

```diff
@@ -342,73 +262,71 @@ class FeelInterpreter {
-      c: (Val, Val, Val) => Boolean,
-      f: Boolean => Val)(implicit context: EvalContext): Val =
+                           x: Val,
+                           y: Val,
+                           c: (Val, Val, Val) => Boolean,
+                           f: Boolean => Val)(implicit context: EvalContext): Val =
     withVal(
       input,
       _ match {
-        case ValNull                             => f(false)
+        case ValNull => f(false)
         case _ if (x == ValNull || y == ValNull) => f(false)
-        case i if (!i.isComparable)              => ValError(s"$i is not comparable")
-        case _ if (!x.isComparable)              => ValError(s"$x is not comparable")
-        case _ if (!y.isComparable)              => ValError(s"$y is not comparable")
-        case i if (i.getClass != x.getClass) =>
-          ValError(s"$i can not be compared to $x")
-        case i if (i.getClass != y.getClass) =>
-          ValError(s"$i can not be compared to $y")
+        case i if (!i.isComparable || !x.isComparable || !y.isComparable) =>
+          error(EvaluationFailureType.NOT_COMPARABLE, s"Can't compare $input with $x and $y")
+        case i if (i.getClass != x.getClass || i.getClass != y.getClass) =>
+          error(EvaluationFailureType.NOT_COMPARABLE, s"Can't compare $input with $x and $y")
         case i => f(c(i, x, y))
       }
     )
 
-  private def withNumbers(x: Val, y: Val, f: (Number, Number) => Val): Val =
+  private def withNumbers(x: Val, y: Val, f: (Number, Number) => Val)(implicit context: EvalContext): Val =
     withNumber(x, x => {
       withNumber(y, y => {
         f(x, y)
       })
     })
 
-  private def withNumber(x: Val, f: Number => Val): Val = x match {
+  private def withNumber(x: Val, f: Number => Val)(implicit context: EvalContext): Val = x match {
     case ValNumber(x) => f(x)
-    case _            => error(x, s"expected Number but found '$x'")
+    case _            => error(EvaluationFailureType.INVALID_TYPE, s"Expected Number but found '$x'")
   }
 
-  private def withBoolean(x: Val, f: Boolean => Val): Val = x match {
+  private def withBoolean(x: Val, f: Boolean => Val)(implicit context: EvalContext): Val = x match {
     case ValBoolean(x) => f(x)
-    case _             => error(x, s"expected Boolean but found '$x'")
+    case _             => error(EvaluationFailureType.INVALID_TYPE,s"Expected Boolean but found '$x'")
   }
 
   private def withBooleanOrNull(x: Val, f: Boolean => Val): Val = x match {
     case ValBoolean(x) => f(x)
     case _             => ValNull
   }
 
-  private def withBooleanOrFalse(x: Val, f: Boolean => Val): Val = x match {
+  private def withBooleanOrFalse(x: Val, f: Boolean => Val)(implicit context: EvalContext): Val = x match {
     case ValBoolean(x) => f(x)
-    case _ => {
-      logger.warn(s"Suppressed failure: expected Boolean but found '$x'")
+    case _ =>
+      error(EvaluationFailureType.INVALID_TYPE, s"Expected Boolean but found '$x'")
```

Comment:
❌ I don't think this should be a failure. The method is called `withBooleanOrFalse` so there already is an expectation that `x` could be of a different type.

Reply:
Good catch. :+1: 

The function `withBooleanOrFalse()` is used in the `if-then-else` evaluation. If the condition is not a boolean, it is useful to report a failure. I will do the same for the function `withBooleanOrNull()`.